Temp file fixes

freddyaboulton opened the pull request:
# Description

Fixes #4063, Fixes #3592

These are the issues I found:
- Saving arrays/PIL images in Gallery/AnnotatedImage components did not respect the DEFAULT_TEMP_DIR variable
- The logic for saving arrays/PIL images in Gallery/AnnotatedImage components did not compute the filename based on a hash of the data like we do for Video/Audio/File etc. This was causing the same file to be saved multiple times with a different filename each time.
- In the temp file creation logic of IOComponent, we would call `NamedTemporaryFile` and then overwrite the name. This would create an empty file every time the method was run. In these methods we already have the desired dirname and filename so we don't need to create a temporary file.

### Repo of first/second point
Run this demo, only three files should be created in the temp dir

![gallery_file_fix](https://github.com/gradio-app/gradio/assets/41651716/a2ab9383-2044-4542-a850-e03a09c35ba1)




### Repo of third point
Running `video_component` demo note that every time you click submit more files are created
![video_temp_file_issue](https://github.com/gradio-app/gradio/assets/41651716/37c28a9e-8861-4444-a19d-73ad617e1fa7)

On this branch, no empty temp files are created
![video_temp_file_fix](https://github.com/gradio-app/gradio/assets/41651716/a341b106-da7b-4256-9435-2d55ea8f47e6)



 


# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have added a short summary of my change to the CHANGELOG.md
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes


# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

abidlabs commented on gradio/components.py:
```diff
@@ -4068,11 +4084,11 @@ def postprocess(
             base_img_path = base_img
             base_img = np.array(_Image.open(base_img))
         elif isinstance(base_img, np.ndarray):
-            base_file = processing_utils.save_array_to_file(base_img)
```
We can remove this method `processing_utils.save_array_to_file()` from `processing_utils.py`

freddyaboulton replied:
Sounds good - its internal but didn't want to delete anything ahead of 4.0

abidlabs commented on gradio/components.py:
```diff
@@ -4068,11 +4084,11 @@ def postprocess(
+            base_file = self.array_to_temp_file(base_img, dir=self.DEFAULT_TEMP_DIR)
+            base_img_path = str(utils.abspath(base_file))
         elif isinstance(base_img, _Image.Image):
-            base_file = processing_utils.save_pil_to_file(base_img)
```
Same, it's no longer used.

abidlabs commented on gradio/components.py:
```diff
@@ -335,6 +338,19 @@ def base64_to_temp_file_if_needed(
         return full_temp_file_path
 
+    def pil_to_temp_file(self, img: _Image.Image, dir: str) -> str:
+        filename = str(Path(dir) / f"{self.hash_bytes(img.tobytes())}.png")
```
Nit: Note that this method creates files with the following filepath:

`<temp_dir>/<hash>.png`

whereas the other methods create filepaths like this:

`<temp_dir>/<hash>/name.suffix`

Maybe it's better to follow the same syntax and do something like 

`<temp_dir>/<hash>/image.png`

because otherwise the filenames are very long when you download the files.

freddyaboulton replied:
Done!

freddyaboulton replied:
This isn't a breaking change but now users will get `audio (13).wav` as opposed to a unique name in their downloads dir

abidlabs commented on gradio/components.py:
```diff
@@ -335,6 +338,19 @@ def base64_to_temp_file_if_needed(
+        return filename
+
+    def array_to_temp_file(self, arr: np.ndarray, dir: str) -> str:
+        filename = str(Path(dir) / f"{self.hash_bytes(arr.tobytes())}.png")
```
Same point as above

abidlabs commented:
Changes look nice @freddyaboulton. Tested with the `video_identity` demo and indeed no extra files are created. However, I think there are still some other components that create temporary files, for example if you test with the `reverse_audio` demo, temporary files still seem to be created in the wrong directory. 

I think the following methods need to be fixed:
* `processing_utils.audio_to_file` (for the `Audio` component)
* `processing_utils.convert_video_to_playable_mp4`  (for the `Video` component)
* `client.utils.decode_base64_to_file` (for the `File` and `UploadButton` components)

That's all the ones I could find scanning the components, but would be good to test with a comprehensive demo like `kitchen_sink` afterwards.

akx commented on gradio/components.py:
```diff
@@ -237,23 +237,28 @@ def hash_url(self, url: str, chunk_num_blocks: int = 128) -> str:
         return sha1.hexdigest()
 
-    def hash_base64(self, base64_encoding: str, chunk_num_blocks: int = 128) -> str:
+    def _hash(self, base64_encoding: str | bytes, chunk_num_blocks: int = 128):
```
The hashing functions should probably be moved into free functions in e.g. `utils.py` (or a new `hashing.py`) as they do not use `self` for anything.

freddyaboulton replied:
Used staticmethod where possible and filed this to track the wider refactor: https://github.com/gradio-app/gradio/issues/4270

akx commented on gradio/components.py:
```diff
@@ -237,23 +237,28 @@ def hash_url(self, url: str, chunk_num_blocks: int = 128) -> str:
+    def _hash(self, base64_encoding: str | bytes, chunk_num_blocks: int = 128):
         sha1 = hashlib.sha1()
         for i in range(0, len(base64_encoding), chunk_num_blocks * sha1.block_size):
             data = base64_encoding[i : i + chunk_num_blocks * sha1.block_size]
-            sha1.update(data.encode("utf-8"))
+            sha1.update(data.encode("utf-8") if isinstance(data, str) else data)
         return sha1.hexdigest()
```
You'd save cycles by letting the hasher do the work if you don't need to encode the data.
```suggestion
        if isinstance(base64_encoding, bytes):
            sha1.update(base64_encoding)
        else:
            for i in range(0, len(base64_encoding), chunk_num_blocks * sha1.block_size):
                data = base64_encoding[i : i + chunk_num_blocks * sha1.block_size]
                sha1.update(data.encode("utf-8"))
        return sha1.hexdigest()

freddyaboulton replied:
Thanks for the suggestion!

freddyaboulton replied:
Thanks for the suggestion. Done!

akx commented on test/test_blocks.py:
```diff
@@ -465,6 +469,57 @@ def test_raise_error_if_event_queued_but_queue_not_enabled(self):
+    def test_pil_images_hashed(self):
+        def create_images(n_images):
+
+            a = Image.new("RGB", (512, 512), "red")
+            b = Image.new("RGB", (512, 512), "green")
+            c = Image.new("RGB", (512, 512), "blue")
```
You might want to move these outside `create_images` to save some cycles.

freddyaboulton replied:
I think it's ok for now in this simple test

akx commented on test/test_blocks.py:
```diff
@@ -465,6 +469,57 @@ def test_raise_error_if_event_queued_but_queue_not_enabled(self):
+    def test_no_empty_files(self):
+        file_dir = pathlib.Path(pathlib.Path(__file__).parent, "test_files")
+        video = str(file_dir / "video_sample.mp4")
+        dir_ = tempfile.mkdtemp()
```
You can use the `tmp_path` Py.test fixture to avoid doing this by hand.

freddyaboulton replied:
Done! Thanks for the suggestion

freddyaboulton commented on client/python/gradio_client/client.py:
```diff
@@ -785,8 +785,8 @@ def serialize(self, *data) -> tuple:
         uploaded_files = self._upload(files)
-        self._add_uploaded_files_to_data(uploaded_files, list(data))
-
+        data = list(data)
```
There was a bug here where the uploaded file info was actually not used in the post request. This fixes it and added corresponding test.

freddyaboulton commented on gradio/components.py:
```diff
@@ -2720,14 +2763,21 @@ def process_single_file(f) -> bytes | tempfile._TemporaryFileWrapper:
+                    self.temp_files.add(path)
+
+                # Creation of tempfiles here
+                file = tempfile.NamedTemporaryFile(
```
Keeping these NamedTemporaryFiles because the docs/api says we return a NamedTemporaryFile. This means that File/UploadComponent will create empty files but I think that's ok until 4.0

freddyaboulton commented:
@abidlabs Thanks for the review! I think I've caught all the temp files so this should be good for another review. This is what the kitchen sink output dir looks like now. I've left a comment explaining why we see some empty files from File components.

<img width="912" alt="image" src="https://github.com/gradio-app/gradio/assets/41651716/5657fe4d-224d-4859-9f8a-ecf24b5bc5b5">

I've addressed all the comments too.

akx commented on gradio/components.py:
```diff
@@ -335,6 +340,41 @@ def base64_to_temp_file_if_needed(
         self.temp_files.add(full_temp_file_path)
         return full_temp_file_path
 
+    def pil_to_temp_file(self, img: _Image.Image, dir: str, format=".png") -> str:
+        temp_dir = Path(dir) / self.hash_bytes(img.tobytes())
+        temp_dir.mkdir(exist_ok=True, parents=True)
+        filename = str(temp_dir / f"image.{format}")
+        img.save(filename, pnginfo=processing_utils.get_pil_metadata(img))
+        return filename
+
+    def img_array_to_temp_file(self, arr: np.ndarray, dir: str) -> str:
+        temp_dir = Path(dir) / self.hash_bytes(arr.tobytes())
+        temp_dir.mkdir(exist_ok=True, parents=True)
+        filename = str(temp_dir / "image.png")
+        pil_image = _Image.fromarray(
+            processing_utils._convert(arr, np.uint8, force_copy=False)
+        )
+        pil_image.save(filename)
+        return filename
+
+    def audio_to_temp_file(
+        self, data: np.ndarray, sample_rate: int, dir: str, format: str
+    ):
+        temp_dir = Path(dir) / self.hash_bytes(data.tobytes())
+        temp_dir.mkdir(exist_ok=True, parents=True)
+        filename = str(temp_dir / f"audio.{format}")
+        processing_utils.audio_to_file(sample_rate, data, filename, format=format)
+        return filename
+
+    def file_bytes_to_file(self, data: bytes, dir: str, file_name: str):
+        path = Path(dir) / self.hash_bytes(data)
+        path.mkdir(exist_ok=True, parents=True)
+        path = path / Path(file_name).name
+        with open(path, "wb") as f:
+            f.write(data)
+            f.flush()
+        return path
```
Maybe add a `# TODO: move to free functions (#4270)` to these new functions so they get moved too when someone tackles #4270?

akx commented on gradio/components.py:
```diff
@@ -335,6 +340,41 @@ def base64_to_temp_file_if_needed(
+    def file_bytes_to_file(self, data: bytes, dir: str, file_name: str):
+        path = Path(dir) / self.hash_bytes(data)
+        path.mkdir(exist_ok=True, parents=True)
+        path = path / Path(file_name).name
+        with open(path, "wb") as f:
+            f.write(data)
+            f.flush()
```
```suggestion
        (path / Path(file_name).name).write_bytes(data)
```

freddyaboulton replied:
Nice

akx commented on gradio/components.py:
```diff
@@ -335,6 +340,41 @@ def base64_to_temp_file_if_needed(
         return full_temp_file_path
 
+    def pil_to_temp_file(self, img: _Image.Image, dir: str, format=".png") -> str:
+        temp_dir = Path(dir) / self.hash_bytes(img.tobytes())
```
Nit: the image bytes don't necessarily describe everything about the image in the presence of e.g. color profiles, metadata, etc.

freddyaboulton replied:
Added a test to make sure we're covered in this case

akx commented on gradio/components.py:
```diff
@@ -335,6 +340,41 @@ def base64_to_temp_file_if_needed(
+        img.save(filename, pnginfo=processing_utils.get_pil_metadata(img))
+        return filename
+
+    def img_array_to_temp_file(self, arr: np.ndarray, dir: str) -> str:
+        temp_dir = Path(dir) / self.hash_bytes(arr.tobytes())
+        temp_dir.mkdir(exist_ok=True, parents=True)
+        filename = str(temp_dir / "image.png")
+        pil_image = _Image.fromarray(
+            processing_utils._convert(arr, np.uint8, force_copy=False)
+        )
+        pil_image.save(filename)
+        return filename
```
Why not
```suggestion
    def img_array_to_temp_file(self, arr: np.ndarray, dir: str, format=".png") -> str:
        pil_image = _Image.fromarray(
            processing_utils._convert(arr, np.uint8, force_copy=False)
        )
        return self.pil_to_temp_file(pil_image, dir=dir, format=format)
```
?
That would ensure a NumPy array describing an image and the equivalent PIL image get the same hash...

freddyaboulton replied:
Good point! Added a test for this too

akx commented on gradio/components.py:
```diff
@@ -1758,12 +1801,11 @@ def _format_image(
-                delete=False,
-                suffix=(f".{fmt.lower()}" if fmt is not None else ".png"),
+            path = self.pil_to_temp_file(
+                im, dir=self.DEFAULT_TEMP_DIR, format=fmt if fmt else ".png"
```
```suggestion
                im, dir=self.DEFAULT_TEMP_DIR, format=fmt or ".png"
```

akx commented on test/test_blocks.py:
```diff
@@ -463,6 +466,115 @@ def test_raise_error_if_event_queued_but_queue_not_enabled(self):
 
+class TestTempFile:
+    def test_pil_images_hashed(self, tmp_path, connect):
+        def create_images(n_images):
+            a = Image.new("RGB", (512, 512), "red")
+            b = Image.new("RGB", (512, 512), "green")
+            c = Image.new("RGB", (512, 512), "blue")
+
+            res = [a, b, c][:n_images]
+            random.shuffle(res)
+            return res
```
> I think it's ok for now in this simple test

I can almost bet money that no one will ever touch this test again, so doing it more efficiently now pays off.

Also, you can use `random.sample()`, so...

```suggestion
                images = [Image.new("RGB", (512, 512), color) for color in ("red", "green", "blue")]
                def create_images(n_images):
                    return random.sample(images, n_images)
```

freddyaboulton replied:
The difference is in microseconds but made the switch. The new code _is_ nicer!

akx commented on test/test_blocks.py:
```diff
@@ -463,6 +466,115 @@ def test_raise_error_if_event_queued_but_queue_not_enabled(self):
+            random.shuffle(res)
+            return res
+
+        with mock.patch.dict(os.environ, {"GRADIO_TEMP_DIR": str(tmp_path)}):
```
Repeated for all of these tests:

Py.test has a `monkeypatch` fixture with `setenv` (which obviously resets the environment back when the fixture dissolves), so you can get by with one less nesting level and shorter tests with

```python
monkeypatch.setenv("GRADIO_TEMP_DIR", str(temp_path))
```
instead of `mock.patch.dict`.

freddyaboulton replied:
Nice! Thanks

akx commented on test/test_processing_utils.py:
```diff
@@ -54,12 +54,12 @@ def test_encode_pil_to_base64(self):
         input_img.info = {"key1": "value1", "key2": "value2"}
 
-        file_obj = processing_utils.save_pil_to_file(input_img)
+        file_obj = components.Image().pil_to_temp_file(input_img, dir=tmp_path)
```
I don't see why these were moved from `processing_utils` to where they need an instance for no reason? Maybe `image_utils`? `file_utils`?

abidlabs replied:
It's to keep all of the functions that create temporary files in one place (specifically in the `IOComponent` class)

abidlabs commented:
LGTM @freddyaboulton great PR! And thanks for all of the suggestions @akx

abidlabs commented:
I'll go ahead and merge this in so that we can work on getting these all these PRs closed :)
